### DISCUSSION TITLE
Question about hls-segment-timeout?

### DISCUSSION THREAD

**Discussion by erickyun** [streamlink/streamlink#3685] [Q&A] [answered]
Can I use these option together?
**--hls-segment-timeout** , **--hls-segment-attempts**

Will hls-segment-attempts option be ignored if I use the two options together?

Example:
So if I set ```--hls-segment-timeout 5 ``` and ```--hls-segment-attempts 3``` will it try to download a segment 3 time after 5 seconds or will it only try 1 time after 5 seconds?

**Comment by bastimeyer** [comment, marked as answer]
> `--hls-segment-attempts ATTEMPTS`
> How many attempts should be done to download each HLS segment before giving up.
> Default is: 3.

https://streamlink.github.io/cli.html#cmdoption-hls-segment-attempts

----

> `--hls-segment-timeout TIMEOUT`
> HLS segment connect and read timeout.
> Default is: 10.0.

https://streamlink.github.io/cli.html#cmdoption-hls-segment-timeout

----

> So if I set --hls-segment-timeout 5 and --hls-segment-attempts 3 will it try to download a segment 3 time after 5 seconds

It'll attempt to download each segment at most 3 times (assuming there was no server error, eg. HTTP status 4xx / 5xx) with each attempt having at most 5 seconds

- https://github.com/streamlink/streamlink/blob/1baa545ee3459564ecfeabcd4c7d9e705d585b24/src/streamlink/stream/hls.py#L121-L123
- https://github.com/streamlink/streamlink/blob/1baa545ee3459564ecfeabcd4c7d9e705d585b24/src/streamlink/plugin/api/http_session.py#L124-L147

**Reply by erickyun** [reply]
Does streamlink wait ```--hls-segment-attempts``` to finish before downloading the next segment or
 does it skip error segment and retry downloading error segments at the end of the download process?

**Reply by bastimeyer** [reply]
Segments are downloaded in parallel, depending on `--hls-segment-threads THREADS` (default is 1), which defines the size of the writer thread's ThreadPoolExecutor, which is responsible for downloading the individual segments. Failed segments mean the end of the stream output, as there would be no coherent output if one segment was missing. There is no "re-queuing" (which would also only make sense for non-live content).

- https://github.com/streamlink/streamlink/blob/1baa545ee3459564ecfeabcd4c7d9e705d585b24/src/streamlink/stream/hls.py#L125-L127
- https://github.com/streamlink/streamlink/blob/1baa545ee3459564ecfeabcd4c7d9e705d585b24/src/streamlink/stream/segmented.py#L170-L178

**Reply by erickyun** [reply]
Sorry for my bad English but I want to be sure I did get it right.

So it would be better if I use a shorter time for ```--hls-segment-timeout``` cuz streamlink would not attempt retry downloading segment at the end but it would attempt retry downloading segment when the error occurs right?

**Reply by bastimeyer** [reply]
`--hls-segment-timeout` limits the time for the segment's HTTP response, until the server has started sending data. If you choose a value that's too low, you will unnecessarily raise errors for HTTP request that would usually be fine. As said, there is no "re-queuing" of failed segments, so if a segment has reached its max download attempts, the stream will end. There's usually no need to change the number of attempts or the timeout value though, as this only affects the connection to the server and its response time. Actual server response errors are not handled by this, so if a segment download fails this way, the stream will end, too.

**Reply by erickyun** [reply]
Thank you very much for your answer I Understand now.
